semantic/check_test.go

```diff
@@ -1216,3 +1338,960 @@ func TestVarAssignmentChecks(t *testing.T) {
+			//
+			name: "dependency across files",
+			pkg: &Package{
+				Package: "foo",
+				Files: []*File{
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Body: []Statement{
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "a"},
+									Init:       &IntegerLiteral{Value: 0},
+								},
+							},
+						},
+					},
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Body: []Statement{
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "b"},
+									Init: &IdentifierExpression{
+										loc: loc{
+											File: "b.flux",
+											Start: ast.Position{
+												Line:   2,
+												Column: 12,
+											},
+											End: ast.Position{
+												Line:   2,
+												Column: 13,
+											},
+										},
+										Name: "a",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: fmt.Errorf(`option dependency: option "b" depends on option "a" defined in the same package at b.flux|2:12-2:13`),
+		},
+		{
+			// package foo
+			// import "bar"
+			// option a = bar.x
+			//
+			name: "dependency on export",
+			pkg: &Package{
+				Package: "foo",
+				Files: []*File{
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Imports: []*ImportDeclaration{
+							{
+								Path: &StringLiteral{Value: "bar"},
+							},
+						},
+						Body: []Statement{
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "a"},
+									Init: &MemberExpression{
+										Object:   &IdentifierExpression{Name: "bar"},
+										Property: "x",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			// package foo
+			// import "bar"
+			//
+			// option a = 0
+			// option a = bar.a.x
+			//
+			name: "option with same name as export",
+			pkg: &Package{
+				Package: "foo",
+				Files: []*File{
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Imports: []*ImportDeclaration{
+							{
+								Path: &StringLiteral{Value: "bar"},
+							},
+						},
+						Body: []Statement{
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "a"},
+									Init:       &IntegerLiteral{Value: 0},
+								},
+							},
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "a"},
+									Init: &MemberExpression{
+										Object: &MemberExpression{
+											Object:   &IdentifierExpression{Name: "bar"},
+											Property: "a",
+										},
+										Property: "x",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			// package foo
+			// option a = 0
+			// option a = a + 1
+			//
```

Comment:
It seems strange that this is not legal but that reassigning `a` with a different constant is.

Reply:
I agree it does seem strange and perhaps this changes in the future if it turns out to be too much of a handicap not allowing any option dependencies. But it is consistent with the definition of options being mutable global variables. And it doesn't introduce any non-determinism so I'd prefer not to restrict reassignments in that way.